compiler/rustc_errors/src/lib.rs

```diff
@@ -1340,9 +1348,22 @@ pub fn add_elided_lifetime_in_path_suggestion(
     );
 }
 
-// Useful type to use with `Result<>` indicate that an error has already
-// been reported to the user, so no need to continue checking.
+/// Useful type to use with `Result<>` indicate that an error has already
+/// been reported to the user, so no need to continue checking.
 #[derive(Clone, Copy, Debug, Encodable, Decodable, Hash, PartialEq, Eq)]
-pub struct ErrorReported;
+pub struct ErrorReported(InnerErrorReported);
+
+/// A secret inner type unconstructable outside `rustc_error`, so that the only way to get an
+/// `ErrorReported` value is by `emit`ing an error to the user.
+#[derive(Clone, Copy, Debug, Encodable, Decodable, Hash, PartialEq, Eq)]
+struct InnerErrorReported;
+
+impl ErrorReported {
+    /// To be used only if you really know what you are doing... ideally, we would find a way to
+    /// eliminate all calls to this method.
+    pub fn new_i_know_what_im_doing() -> Self {
```

Comment:
Can you use a more self-descriptive name? Like `unchecked_claim_error_was_emitted` or something similar.

We could have it set a `thread_local!` that panics if not defused by some diagnostic emission on the same thread or something, but it might not be worth it if calls can be removed.

Reply:
> We could have it set a thread_local! that panics if not defused by some diagnostic emission on the same thread or something, but it might not be worth it if calls can be removed.

I'll add a note in the tracking issue for the future.